components/automate-ui/src/app/services/chef-session/chef-session.service.ts

```diff
@@ -61,36 +70,45 @@ export class ChefSessionService implements CanActivate {
       // and components/session-service and chef-session-service -- it would
       // make sense to try to refresh the session if we get a 401 from the API,
       // before giving up and calling logout().
-      window.setInterval(() => {
-        /* Using XMLHttpRequest instead of angular/http to avoid
-           dependency hell. */
-        const xhr = new XMLHttpRequest();
-        xhr.onreadystatechange = this.refreshSessionCallback.bind(this);
-        xhr.open('GET', '/session/refresh', true);
-        xhr.responseType = 'json';
-        xhr.setRequestHeader('Authorization', `Bearer ${this.id_token}`);
-        xhr.send(null);
-      }, minute);
+      timer(0, minute).pipe(
+        filter(_ => !this.isRefreshing),
+        mergeMap(_ => {
+          this.isRefreshing = true;
+          return this.refresh();
+        }),
+      ).subscribe(
+        token => {
+          this.ingestIDToken(token);
+          this.isRefreshing = false;
+        },
+        error => {
+          this.isRefreshing = false;
+          if (error instanceof HttpErrorResponse) {
+            if (error.status == HTTP_STATUS_UNAUTHORIZED) {
+              this.logout();
+            } else {
+              console.log(`Session refresh failed: ${error.statusText}`);
+            }
+          } else {
+            console.log(error);
+          }
+        }
+      );
     }
-
-    this.tryInitializeSession();
   }
 
-  refreshSessionCallback(event): void {
-    const xhr = <XMLHttpRequest>event.target;
-    if (xhr.readyState === XHR_DONE) {
-      if (xhr.status === HTTP_STATUS_OK) {
-        this.ingestIDToken(xhr.response.id_token);
-      } else if (xhr.status === HTTP_STATUS_UNAUTHORIZED) {
-        this.logout();
-      } else {
-        // TODO 2017/12/15 (sr): is there anything we could do that's better than
-        // this?
-        // NOTE 2019/06/27 (sr): this defaults to OK when using HTTP2; not worth
-        // messing with, though, this all needs to be redone.
-        console.log(`Session refresh failed: ${xhr.statusText}`);
-      }
-    }
+  private refresh(): Observable<string> {
+    const httpOptions = {
+      headers: new HttpHeaders({
+        'Content-Type': 'application/json',
+        'Authorization': `Bearer ${this.id_token}`
+      })
+    };
+    return this.httpHandler.get('/session/refresh', httpOptions).pipe(
+      map(obj => {
+        return obj['id_token'];
+      }),
```

Comment:
[nit]
```suggestion
      map(obj => obj.id_token),
```

would do it, too, I think

Reply:
i feel like i tried this and something was complaining

Reply:
Type issues?